src/typing/env_js.ml

```diff
@@ -40,19 +40,22 @@ open Scope
    2. ForValue lookups give errors if they forward reference non-hoisted
       things (lets or consts)
 
-   3. ForType lookups may return values or type aliases, since some values
+   3. ForUpdate lookups give errors in all cases ForValue does, with the
+      additional constraint that the found entry must not be const.
+
+   4. ForType lookups may return values or type aliases, since some values
       also denote types - e.g. a generator function F also denotes the type
       of the objects it creates. Of course many values don't also have a type
       denotation and thus errors in type position. But we don't know the type
       of a symbol during local inference as a rule, so errors of this kind are
       not raised here.
 
-   4. ForTypeof lookups are in fact ForValue lookups, but due to the order in
+   5. ForTypeof lookups are in fact ForValue lookups, but due to the order in
       which AST traversal takes place, these lookups may legitimately violate
       rule #2, hence the need for a special mode.
  *)
 module LookupMode = struct
-  type t = ForValue | ForType | ForTypeof
+  type t = ForValue | ForUpdate | ForType | ForTypeof
```

Comment:
OK, so I went down this path and things got pretty ugly. Happy to go over it in more depth, but I think the alternative approach of defining a set_var-like entrypoint for Update/PlusAssign. As far as a name for that entrypoint... any ideas? To be honest, I don't really see the problem with ForUpdate, so maybe this is a better change for you to make separately/subsequently?